### DISCUSSION TITLE
GEOSException: IllegalArgumentException: Invalid number of points in LinearRing found 2 - must be 0 or >= 4

### DISCUSSION THREAD

**Discussion by EwoutH** [geopandas/geopandas#3336] [General]
I'm trying to load in a shapefile with GeoPandas, but with both supported engines it currently fails.

To reproduce:
```Python
import geopandas as gpd
```

Extract this shapefile: [GIS-VMRDH-Areas.zip](https://github.com/user-attachments/files/15760076/GIS-VMRDH-Areas.zip)

Try to run either with [pyogrio](https://github.com/geopandas/pyogrio) or [fiona](https://github.com/Toblerity/Fiona) engine:

```Python
# Load GIS-VMRDH-Areas with pyogrio engine
areas = gpd.read_file('GIS-VMRDH-Areas/areas_v2.shp', engine="pyogrio")
```
```
GEOSException: IllegalArgumentException: Invalid number of points in LinearRing found 2 - must be 0 or >= 4
```
and
```Python
# Load GIS-VMRDH-Areas with pyogrio engine
areas = gpd.read_file('GIS-VMRDH-Areas/areas_v2.shp', engine="pyogrio")
```
```
ValueError: A linearring requires at least 4 coordinates.
```

Uploading the shapefiles to something like https://mapshaper.org/ or using ArcGIS they are loading fine, without trouble.

I don't think this is a GeoPandas bug, but since it's happening with multiple engines and not with other tools, I'm wondering where the bug is actually happening.

**Comment by brendan-ward** [comment]
The shapefile appears to have a number of issues according to GDAL; in pyogrio it raises a number of warnings when reading the file (there are 29 invalid geometries):

```
raw.py:196: RuntimeWarning: GIS-VMRDH-Areas/areas_v2.shp contains polygon(s) with rings with invalid winding order. Autocorrecting them, but that shapefile should be corrected using ogr2ogr for example.
  return ogr_read(
raw.py:196: RuntimeWarning: organizePolygons() received an unexpected geometry.  Either a polygon with interior rings, or a polygon with less than 4 points, or a non-Polygon geometry.  Return arguments as a collection.
  return ogr_read(
raw.py:196: RuntimeWarning: Geometry of polygon of fid 594 cannot be translated to Simple Geometry. All polygons will be contained in a multipolygon.
...
```

The actual failure happens when the WKB form of the geometry is attempted to be converted into Shapely geometry objects.
The Shapely [from_wkb](https://shapely.readthedocs.io/en/stable/reference/shapely.from_wkb.html#shapely.from_wkb) provides additional options to warn or ignore these errors, but in both cases the invalid geometries are set to `None` instead of a geometry object.  One workaround would be to do something similar to pyogrio's [internal approach for creating a GeoDataFrame](https://github.com/geopandas/pyogrio/blob/main/pyogrio/geopandas.py#L302-L320), and add in the `on_invalid='warn'` (or ignore) in the call to `from_wkb` (not sure we want to actually add this capability to pyogrio, but I guess that is open to discussion).

Is it possible to fix these issues in whatever process produced the shapefile here?  Given the invalid geometries would be dropped by any workarounds we can offer here, it would be more ideal to make sure that the shapefile is written with valid geometries instead so you don't lose any data.

**Reply by martinfleis** [reply]
> add in the on_invalid='warn' (or ignore)

I'd vote for adding this to read_dataframe. I think that this is what QGIS is implicitly doing and I tried to looked for a GDAL flag or anything that would at least allow reading incomplete data but haven't found anything.

**Comment by theroggy** [comment]
Issue has been raised (and discussed) before in #2791...

**Reply by martinfleis** [reply]
My memory is getting wonky... at least my suggestions are consistent :D.

**Comment by EwoutH** [comment]
Thanks for getting back. Unfortunately, I don't have access to the source geometry, I got this Shapefile delivered as-is. Since it shows up correctly in a tool like https://mapshaper.org/ there should be a way to load these in correctly, even with some LinearRings only having 2 points (or maybe it's wrongly detected as LinearRing?).

![image](https://github.com/geopandas/geopandas/assets/15776622/735fb70e-5e84-47be-95c6-85b81f6d368d)
